### DISCUSSION TITLE
Dao Helpers

### DISCUSSION THREAD

**Discussion by hlanderdev** [pocketbase/pocketbase#818] [Q&A] [answered]
Hi, it was mentioned in discussion #734 that through creating your own models with the models.Model interface, you could write your own dao helpers. I was just wondering if there were any more in-depth examples of that. I am fairly new to Go, but I know that it does not have inheritance. The closest thing seems to be embedded structs, but I couldn't quite get that to work. Or was the point of that statement about writing separate helper functions and not necessarily extending the base Dao itself? Thanks.

**Comment by ganigeorgiev** [comment, marked as answer]
@highlanderdev More detailed examples will be available soon after the v0.8 pre-release.

>  The closest thing seems to be embedded structs, but I couldn't quite get that to work.

It depends on what were you trying to embed. The example in #734 embeds `models.BaseModel`.

> Or was the point of that statement about writing separate helper functions and not necessarily extending the base Dao itself?

Yes, you can't change external package definitions. What I meant was to create your own subpackage where each helper function would accept `app.Dao()` instance and use the injected `Dao` with your custom model struct. Or in other words, instead of having:

```go
func (dao *daos.Dao) MyHelper(...someArgs...) {}
```
you'll have something like:
```go
func MyHelper(dao *daos.Dao, ...someArgs...) {}
```

Of course, you can always embed a `daos.Dao` struct in your own custom Dao instance, but then you'll not be able to make use of it using `app.Dao()` and you'll have to wrap it with your custom struct, eg. `MyCustomDao{app.Dao()}` every time when you want to initialize it (altenatively you could store it as a global singleton somewhere).

**Reply by hlanderdev** [reply]
Thanks! This answers my question. I think I was just over-thinking the issue. The injection solution seems more preferable than using a singleton with an embedded Dao.
